client/app/bundles/course/assessment/question/programming/ProgrammingForm.tsx

```diff
@@ -0,0 +1,154 @@
+import loadingToast from './hooks/loadingToast';
+import { ProgrammingFormDataProvider } from './hooks/ProgrammingFormDataContext';
+import useLanguageMode from './hooks/useLanguageMode';
+import { watchEvaluation } from './operations';
+
+interface ProgrammingFormProps {
+  with: ProgrammingFormData;
+  dirty?: boolean;
+  onSubmit?: (data: ProgrammingFormData) => Promise<ProgrammingPostStatusData>;
+  revalidate?: (
+    response: ProgrammingPostStatusData,
+    data: ProgrammingFormData,
+  ) => Promise<ProgrammingFormData>;
+}
+
+const ProgrammingForm = (props: ProgrammingFormProps): JSX.Element => {
+  const [data, setData] = useState(props.with);
+
+  const { t } = useTranslation();
+
+  const [submitting, setSubmitting] = useState(false);
+  const [form, setForm] = useState<FormEmitter<ProgrammingFormData>>();
+  const [pending, setPending] = useState<() => void>();
+
+  const { languageOptions, getModeFromId } = useLanguageMode(data.languages);
+
+  const submitForm = async (rawData: ProgrammingFormData): Promise<void> => {
+    if (!props.onSubmit) return undefined;
+
+    setSubmitting(true);
+
+    const { update, success, error } = loadingToast(
+      t(translations.savingChanges),
+    );
+
+    try {
+      const response = await props.onSubmit(rawData);
+
+      const toastSuccessAndRedirect = (): void => {
+        success(t(translations.questionSavedRedirecting));
+        window.location.href = response.redirectAssessmentUrl;
+      };
+
+      if (!response.importJobUrl) return toastSuccessAndRedirect();
+
+      update(t(translations.evaluatingSubmissions));
+
+      let debounced = false;
+
+      return watchEvaluation(
+        response.importJobUrl,
+        () => {
+          if (debounced) return;
+          debounced = true;
+
+          toastSuccessAndRedirect();
+        },
+        async () => {
+          if (debounced) return;
+          debounced = true;
+
+          const newData = await props.revalidate?.(response, rawData);
+          if (newData) {
+            setData(newData);
+            form?.resetTo?.(newData);
+          }
+
+          error(t(translations.questionSavedButPackageError));
+
+          setSubmitting(false);
+          window.location.href = `#${PACKAGE_SECTION_ID}`;
+        },
+      );
+    } catch {
+      setSubmitting(false);
+      return error(t(translations.errorWhenSavingQuestion));
+    }
+  };
+
+  const preProcessForm = (draft: ProgrammingFormData): ProgrammingFormData => {
+    if (draft.testUi?.mode) {
+      draft.testUi.mode = getModeFromId(draft.question.languageId);
+    }
+
+    return draft;
+  };
+
+  return (
+    <ProgrammingFormDataProvider from={data}>
+      <Form
+        contextual
+        dirty={props.dirty}
+        disabled={submitting}
+        emitsVia={setForm}
+        headsUp
+        initialValues={data}
+        onSubmit={(rawData): void => {
+          if (data.question.shouldWarnOnSubmit) {
```

Comment:
Could you also checked if any code related field (i.e. prepend, append, etc), updated attachment, updated time/mem limit is dirty?